nixos/modules/services/web-apps/keycloak.nix

```diff
@@ -268,67 +264,186 @@ in
         '';
       };
 
-      extraConfig = mkOption {
-        type = attrsOf anything;
-        default = { };
+      settings = mkOption {
+        type = lib.types.submodule {
+          freeformType = attrsOf (nullOr (oneOf [ str int bool (attrsOf path) ]));
+
+          options = {
+            http-host = mkOption {
+              type = str;
+              default = "0.0.0.0";
+              example = "127.0.0.1";
+              description = ''
+                On which address Keycloak should accept new connections.
+
+                A special syntax can be used to allow command line Java system
+                properties to override the value: ''${property.name:value}
+              '';
+            };
+
+            http-port = mkOption {
+              type = port;
+              default = 80;
+              example = 8080;
+              description = ''
+                On which port Keycloak should listen for new HTTP connections.
+
+                A special syntax can be used to allow command line Java system
+                properties to override the value: ''${property.name:value}
+              '';
+            };
+
+            https-port = mkOption {
+              type = port;
+              default = 443;
+              example = 8443;
+              description = ''
+                On which port Keycloak should listen for new HTTPS connections.
+
+                A special syntax can be used to allow command line Java system
+                properties to override the value: ''${property.name:value}
+              '';
+            };
+
+            http-relative-path = mkOption {
+              type = str;
+              default = if versionAtLeast config.system.stateVersion "22.05" then "" else "/auth";
```

Comment:
This eases migration for existing users, but may be surprising to those looking at the new keycloak docs.

Reply:
Yeah, I suppose the alternative would be to just make it very clear that the user must make the choice. Since the `frontendUrl` option is removed, which will prompt users to look at this option, maybe that's preferable.

Reply:
I've removed this now and made the error message more elaborate. With that and a notice in the breaking changes section of the release notes, I think we should be good.